### PR TITLE
Adding the guide for using RisingWave UDF API for Python

### LINKED ISSUE #659: Document UDF
### Related code PR

https://github.com/risingwavelabs/risingwave/pull/8597
https://github.com/risingwavelabs/risingwave/pull/8255

### Which part(s) of the docs might be affected or should be updated? And how?

We need a dedicated guide about UDF.

### Reference

_No response_

### PR DESCRIPTION
**Pull request by CharlieSYH**
## Info

- **Preview**: 
https://pr-689.d2fbku9n2b6wde.amplifyapp.com/docs/upcoming/user-defined-functions/

- **Related code PR**: 
https://github.com/risingwavelabs/risingwave/pull/8597

- **Related doc issue**: 
Resolves https://github.com/risingwavelabs/risingwave-docs/issues/659

- **Notes**: 
[ Any additional information? ]

## Before merging
  - [ ] (For version-specific PR) I have selected the corresponding software version in **Milestone** and linked the related doc issue to this PR in **Development**.
  - [ ] I have acquired the approval from the owner (and optionally the reviewers) of the code PR and at least one tech writer (`bernscode`, `CharlieSYH`, `emile-00`, & `hengm3467`). 
  - [ ] I have checked the doc site preview, and the updated parts look good. <details><summary>How?</summary>Scroll down and open this link: image
</details>

### CODE REVIEW

File: docs/sql/query-syntax/user-defined-functions.md [hunk 1]
```diff
@@ -0,0 +1,225 @@
...
+
+# Start a UDF server
+if __name__ == '__main__':
+    server = UdfServer(location="0.0.0.0:8815")
+    server.add_function(gcd)
+    server.add_function(series)
+    server.serve()
+```
+<details>
+<summary>See code explanation</summary>
```
Comment (CharlieSYH) [inline docs/sql/query-syntax/user-defined-functions.md:114]: @wangrunji0408 Please review the code explanation.

File: docs/sql/query-syntax/user-defined-functions.md [hunk 2]
```diff
@@ -0,0 +1,225 @@
...
+    USING LINK 'udf_server_address';
+```
+
+</TabItem>
+
+</Tabs>
+
+
+
+| Parameter or clause | Description |
```
Comment (CharlieSYH) [inline docs/sql/query-syntax/user-defined-functions.md:216]: @wangrunji0408 Please review the parameter descriptions.

File: docs/sql/query-syntax/user-defined-functions.md [hunk 3]
```diff
@@ -0,0 +1,225 @@
...
+
+```sql
+CREATE FUNCTION gcd(int, int) RETURNS int
+LANGUAGE python AS gcd USING LINK 'http://localhost:8815';
+
+CREATE FUNCTION series(int) RETURNS TABLE (x int)
+LANGUAGE python AS series USING LINK 'http://localhost:8815';
+```
+
+## 5. Use your functions in RisingWave
```
Comment (CharlieSYH) [inline docs/sql/query-syntax/user-defined-functions.md:242]: @wangrunji0408 Can you provide known limitations for creating/using UDFs in RIsingWave at this stage?

File: sidebars.js [hunk 4]
```diff
@@ -310,6 +310,10 @@ const sidebars = {
             id: 'sql/query-syntax/query-syntax-with-clause',
             label: 'WITH',
           },
+          {
```
Comment (CharlieSYH) [inline sidebars.js:313]: @hengm3467 Should I add (a link to) CREATE FUNCTION to the sidebar under SQL - Commands?

File: docs/sql/query-syntax/user-defined-functions.md [hunk 5]
```diff
@@ -0,0 +1,225 @@
...
+---
+
+You can define your own functions (including table functions) and call these functions in RisingWave. With the user-defined function (UDF), you can tailor RisingWave to your needs and take advantage of the power and flexibility of Python to create complex and customized functionality for your data processing and analysis tasks.
+Currently, RisingWave supports UDFs implemented as external functions in Python.
+
+This article provides a step-by-step guide for installing the RisingWave UDF API, defining functions in a Python file, starting the UDF server, and declaring and using UDFs in RisingWave.
+
+## Prerequisites
+
+- Ensure that you have [Python](https://www.python.org/downloads/) installed on your computer.
```
Comment (wangrunji0408) [inline docs/sql/query-syntax/user-defined-functions.md]: Please mention that the minimum supported version is Python 3.8.

File: docs/sql/query-syntax/user-defined-functions.md (hunk 1 above)
Comment (wangrunji0408) [reply docs/sql/query-syntax/user-defined-functions.md:114]: LGTM!

File: docs/sql/query-syntax/user-defined-functions.md [hunk 6]
```diff
@@ -0,0 +1,225 @@
...
+</Tabs>
+
+
+
+| Parameter or clause | Description |
+| --- | --- |
+| *function_name* | The name of the UDF that you want to declare in RisingWave. |
+| *argument_type* | The data type of the input parameter(s) that the UDF expects to receive.|
+| **RETURNS** *return_type* | Specifies the data type of the return value from the UDF. |
+| **RETURNS TABLE** | Specifies the structure of the table that the UDF returns. |
```
Comment (wangrunji0408) [inline docs/sql/query-syntax/user-defined-functions.md]: Please mention that the former is for scalar functions and the latter is for table functions. User should use either of them.

File: docs/sql/query-syntax/user-defined-functions.md [hunk 7]
```diff
@@ -0,0 +1,225 @@
...
+CREATE FUNCTION gcd(int, int) RETURNS int
+LANGUAGE python AS gcd USING LINK 'http://localhost:8815';
+
+CREATE FUNCTION series(int) RETURNS TABLE (x int)
+LANGUAGE python AS series USING LINK 'http://localhost:8815';
+```
+
+## 5. Use your functions in RisingWave
+
+After you have declared your UFDs in RisingWave, you can use them in SQL queries just like any built-in functions.
```
Comment (wangrunji0408) [inline docs/sql/query-syntax/user-defined-functions.md]:
```suggestion
After you have declared your UDFs in RisingWave, you can use them in SQL queries just like any built-in functions.
```

File: docs/sql/query-syntax/user-defined-functions.md (hunk 3 above)
Comment (wangrunji0408) [reply docs/sql/query-syntax/user-defined-functions.md:242]: 1. `ARRAY` and `JSONB` types are not supported now. We will add them soon.
2. For `STRUCT` types, field names must be consistent between Python and SQL definitions. Otherwise it will be considered a type mismatch. This requirement may be relaxed in the future.

File: docs/sql/query-syntax/user-defined-functions.md [hunk 8]
```diff
@@ -0,0 +1,225 @@
...
+    while y != 0:
+        (x, y) = (y, x % y)
+    return x
+
+# Define a table function
+@udtf(input_types='INT', result_types='INT')
+def series(n):
+    for i in range(n):
+        yield i
+
```
Comment (wangrunji0408) [inline docs/sql/query-syntax/user-defined-functions.md:104]: We can add an example of returning multiple values.

```python
# Define a scalar function that returns multiple values
@udf(input_types=['BINARY'], result_type='STRUCT<src_ip VARCHAR, dst_ip VARCHAR, src_port SMALLINT, dst_port SMALLINT>')
def extract_tcp_info(tcp_packet: bytes):
    src_addr, dst_addr = struct.unpack('!4s4s', tcp_packet[12:20])
    src_port, dst_port = struct.unpack('!HH', tcp_packet[20:24])
    src_addr = socket.inet_ntoa(src_addr)
    dst_addr = socket.inet_ntoa(dst_addr)
    return src_addr, dst_addr, src_port, dst_port
```

```sql
create function extract_tcp_info(bytea) returns struct<src_ip varchar, dst_ip varchar, src_port smallint, dst_port smallint>
language python as extract_tcp_info using link 'http://localhost:8815';

select extract_tcp_info(E'\\x45000034a8a8400040065b8ac0a8000ec0a80001035d20b6d971b900000000080020200493310000020405b4' :: bytea);
(192.168.0.14,192.168.0.1,861,8374)
```

Comment (CharlieSYH) [reply docs/sql/query-syntax/user-defined-functions.md:104]: Thanks! All comments above have been resolved.

**Comment by CharlieSYH** [comment]
This is still a WIP. Need to think about how to use UDFs when using Docker to run RisingWave.

File: sidebars.js (hunk 4 above)
Comment (hengm3467) [reply sidebars.js:313]: @CharlieSYH Yes. Let's do that. And point to the new guide. This ensures findability via both self-navigation and search.

File: docs/sql/query-syntax/user-defined-functions.md [hunk 9]
```diff
@@ -0,0 +1,269 @@
+---
+id: user-defined-functions
+slug: /user-defined-functions
+title: User-defined functions
+description: Define your own functions with the help of the RisingWave UDF API for Python.
+---
+
+You can define your own functions (including table functions) and call these functions in RisingWave. With the user-defined function (UDF), you can tailor RisingWave to your needs and take advantage of the power and flexibility of Python to create complex and customized functionality for your data processing and analysis tasks.
```
Comment (hengm3467) [inline docs/sql/query-syntax/user-defined-functions.md]: The second paragraph is a bit too wordy. Suggestion: ...-> ..of Python to perform complex and customized data processing and analysis tasks.

File: docs/sql/query-syntax/user-defined-functions.md [hunk 10]
```diff
@@ -0,0 +1,269 @@
...
+<details>
+<summary>Cannot run this command?</summary>
+If "command not found: pip" is returned, <a href="https://packaging.python.org/en/latest/tutorials/installing-packages/#ensure-you-can-run-pip-from-the-command-line">check if pip is available</a> in your environment and <a href="https://packaging.python.org/en/latest/tutorials/installing-packages/#ensure-pip-setuptools-and-wheel-are-up-to-date">ensure it is up to date</a>.
+</details>
+
+
+
+## 2. Define your functions in a Python file
+
+To better demonstrate this step, we have prepared a sample script for you to try out. Please create a Python file with the name "udf.py" and insert the script below.
```
Comment (hengm3467) [inline docs/sql/query-syntax/user-defined-functions.md]: I think file names should also be wrapped in code blocks. Correct me if I'm wrong.

File: docs/sql/query-syntax/user-defined-functions.md [hunk 11]
```diff
@@ -0,0 +1,269 @@
...
+| **RETURNS TABLE** | Use this if the function is a table-valued function (TVF). It specifies the structure of the table that the UDF returns. |
+| **LANGUAGE** | Specifies the programming language used to implement the UDF. <br/> Currently, only `python` is supported.|
+| **AS** *function_name_defined_in_server* | Specifies the function name defined in the UDF server.|
+| **USING LINK** '*udf_server_address*' | Specifies the server address where the UDF implementation resides. <br/>If you are running RisingWave in your local environment, the address is `http://localhost:<port>` <br/> If you are running RisingWave using Docker, the address is `http://host.docker.internal:<port>/`|
+
+#### Example
+
+Here's the SQL statements for declaring the three UDFs defined in [step 2](#2-define-your-functions-in-a-python-file).
```
Comment (hengm3467) [inline docs/sql/query-syntax/user-defined-functions.md]: Here's -> Here are...

**Review by hengm3467 — APPROVED** [review]
Left some comments. Otherwise looks good to me. Really solid work. Thanks!
